Added QCDregions and subcatogories

Sarang-cern opened the pull request:
1. We introduced the jet‐veto subcategories not because the core physics changed, but to work around a known data/MC mismatch in a narrow η–pₜ window of our forward jets. In the region 2.5 < |η| < 3.0 at moderate pₜ (30–50 GeV), the simulation showed unphysical spikes or horns that didn’t appear in data (https://gitlab.cern.ch/cms-jetmet/coordination/coordination/-/issues/113), driven by imperfect jet energy corrections and resolution(and under further investigation). Those mismodeled jets can skew our background shapes or pull the fit in the VBF and ggH categories, compromising sensitivity. We define a veto requiring each leading jet to be either truly central (|η|<2.5), very forward (|η|>3), or above 50 GeV. Subcategories added are: VBF_JetVeto  (VBF events that pass this extra clean-jet requirement) and ggH_JetVeto (baseline events that fail the VBF_JetVeto_def). This is a temporary fix in the current pull request. 
2. We split the dimuon mass range into four simple slices (calling them QCDregions for now) so we can check backgrounds and keep the Higgs fit clean, we keep the same naming convention used in the previous analysis. Together, these regions give us control over backgrounds and lets us extract the Higgs signal without mixing in data where we are fitting.

valeriadamante reviewed (changes requested):
Please apply the comments

Sarang-cern commented:
Comments have been addressed  for global.yaml, also in H_mumu.py a loop has been adjusted as following:
Unchanged things: channel loop, trigger lookup, category loop, key/expr building.
Changed things: Replaced the old sign_regions_to_consider list with the region_defs dict.
                             Iterated for reg_name, reg_cut in region_defs.items():
                             Used reg_name in the key tuple and reg_cut in the filter expression.

valeriadamante commented on Analysis/H_mumu.py:
```diff
@@ -14,12 +14,12 @@
     filter_str = ""
     channels_to_consider = global_cfg_dict['channels_to_consider']
-    sign_regions_to_consider = global_cfg_dict['QCDRegions']
+    sign_regions_to_consider = global_cfg_dict['QCDregions']
```
This has to be the same as written in `config/global.yaml`. Thus `'QCDRegions'`, not `'QCDregions'`. Check also in flaf how is it written and uniform it within Hmumu framework

Sarang-cern replied:
Done

Sarang-cern replied:
I checked in HistMerger, it is written as "QCDregions" so I have kept it with the small "r".

valeriadamante replied:
The important is that it is uniform everywhere

valeriadamante replied:
+1

valeriadamante commented on Analysis/H_mumu.py:
```diff
@@ -35,6 +35,7 @@ def createKeyFilterDict(global_cfg_dict, year):
                 filter_dict[key] = filter_str
     return filter_dict
 
+
```
remove

Sarang-cern replied:
Can you elaborate more what exactly is to be removed here?

valeriadamante replied:
Did you apply this?

valeriadamante commented on Analysis/H_mumu.py:
```diff
@@ -201,8 +202,15 @@ def defineTriggers(self):
+    for reg_name, reg_cut in region_defs.items():
+        # reg_name → branch name, reg_cut → actual selection
+        self.df = self.df.Define(reg_name, reg_cut)
+    return self.df
```
Did you check how histograms are defined?

Sarang-cern replied:
Nope, can you give me an exact reference? I created this loop according to the QCDRegions dict so that it would pull the keys and values and apply the cuts. Does this loop look correct to you?

valeriadamante commented on config/global.yaml:
```diff
@@ -52,30 +52,19 @@ categories:
+  VBF_def: "HasVBF && j1_pt > 25 && j2_pt > 25"
   VBF: "baseline && VBF_def"
   ggH: "baseline && !(VBF_def)"
-  # https://gitlab.cern.ch/cms-jetmet/coordination/coordination/-/issues/113 --> data/MC differences in Jet distributions
```
please keep this comment as it contains relevant references

Sarang-cern replied:
Yes, just added those comments in my globalyaml and pushing it now

valeriadamante reviewed (changes requested):
Apply these asap

valeriadamante commented on Analysis/H_mumu.py:
```diff
@@ -201,8 +202,12 @@ def defineTriggers(self):
+        for reg_name, reg_cut in region_defs.items():
+            self.df = self.df.Define(reg_name, reg_cut)
+        return self.df
 
```
You don't need to return the df.

valeriadamante commented on config/global.yaml:
```diff
@@ -35,11 +34,11 @@ corrections:
+  Signal_Fit: "m_mumu > 115 && m_mumu < 135"
+  H_sideband: "((m_mumu > 110 && m_mumu < 115) || (m_mumu > 135 && m_mumu < 150))"
+  Signal_ext: "m_mumu > 110 && m_mumu < 150"
 
```
+1

valeriadamante commented on config/global.yaml:
```diff
@@ -52,30 +51,17 @@ categories:
-  baseline: "OS && (mu1_pt > {MuPtTh} && mu2_pt > {MuPtTh})" # OS = opposite charge sign for two muons
-  VBF_def : "HasVBF && j1_pt > 25 && j2_pt >25"
+  baseline: "OS && (mu1_pt > {MuPtTh} && mu2_pt > {MuPtTh})"
+  VBF_def: "HasVBF && j1_pt > 25 && j2_pt > 25"
```
+1

valeriadamante commented on config/global.yaml:
```diff
@@ -52,30 +51,17 @@ categories:
-
+  VBF_JetVeto_def: "VBF_def && ((abs(j1_eta) < 2.5 || abs(j1_eta) > 3 || j1_pt > 50) && (abs(j2_eta) < 2.5 || abs(j2_eta) > 3 || j2_pt > 50))"
+  VBF_JetVeto: "baseline && VBF_JetVeto_def"
+  ggH_JetVeto: "baseline && !VBF_JetVeto_def"
```
+1

valeriadamante commented on config/global.yaml:
```diff
@@ -99,4 +83,4 @@ sample_types_to_merge:
   # - ST
-  # - TTVV
\ No newline at end of file
+  # - TTVV
```
+1

Sarang-cern commented on Analysis/H_mumu.py:
```diff
@@ -18,7 +18,7 @@
     filter_dict = {}
     filter_str = ""
-    channels_to_consider = global_cfg_dict['channels_to_consider']
+    channels_to_consider = global_cfg_dict['channelSelection']
```
Previously when I was running the script, 2 days ago I guess, it threw an error saying channels_to_consider unknown line, so then I went to global.yaml and checked how we refer the channels into the H_mumu.py, and once I put the same keyword i.e. channelSelction, the error went away.

Sarang-cern replied:
This was the error I was getting, which went away after I replaced channels_to_consider with channelSelection.

Traceback (most recent call last):
  File "/afs/cern.ch/user/s/sabhosal/H_mumu/FLAF/Analysis/HistProducerFile.py", line 273, in <module>
    dfWrapped_central = analysis.DataFrameBuilderForHistograms(
        ROOT.RDataFrame('Events', args.inFile),
        global_cfg_dict, 
        args.period, 
        **kwargset
    ).df
  File "/afs/cern.ch/user/s/sabhosal/H_mumu/Analysis/H_mumu.py", line 15, in createKeyFilterDict
    channels_to_consider = global_cfg_dict['channels_to_consider']
KeyError: 'channels_to_consider'

valeriadamante replied:
Because when you run HistProducerFile you have to give the channels as input

valeriadamante replied:
so please come back to the "channels_to_consider"

Sarang-cern replied:
Done and pushed

valeriadamante commented on Analysis/make_stackplot.py:
```diff
@@ -1,45 +1,80 @@
+                        f"--analysis H_mumu "
+                        f"--qcdregion {qcd_region} "
+                        f"--sigConfig ../config/{era}/samples.yaml"
+                    )
```
Did you apply this?
